Rename object files so that we don't have collisions between different

NlightNFotis opened the pull request:
Rename object files so that we don't have collisions between different object files coming from different projects.

Fixes #7586

Co-authored-by: Enrico Steffinlongo "enrico.steffinlongo@diffblue.com"
Co-authored-by: Fotis Koutoulakis "fotis.koutoulakis@diffblue.com"

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

NlightNFotis committed:
Rename object files so that we don't have collisions between different
object files coming from different projects.

Fixes #7586

Co-authored-by: Enrico Steffinlongo "enrico.steffinlongo@diffblue.com"
Co-authored-by: Fotis Koutoulakis "fotis.koutoulakis@diffblue.com"

tautschnig commented on src/libcprover-cpp/add_dependencies.sh:
```diff
@@ -15,11 +15,36 @@ DESTINATION=$1
   # Unpack the library
   ${AR_COMMAND} -x ${lib}
+
+  # Rename all object file in the library prepending "${LIBNAME}_" to avoid
```
Nit pick: s/file/files/

tautschnig commented on src/libcprover-cpp/add_dependencies.sh:
```diff
@@ -15,11 +15,36 @@ DESTINATION=$1
 
 # Append all the unpacked files to the destination library
-${AR_COMMAND} -rcs ${DESTINATION} *.o
+${AR_COMMAND} -rcs ${DESTINATION} **/*.o
```
Why `**`?

thomasspriggs replied:
`**` supports matching the pattern across the contents of multiple directories, which is needed because of the preceding changes which places the `.o` files in separate directories.

esteffin replied:
Because the files are in the subfolder we need to add all `.o` in the subfolders.
However this can be problematic, so Fotis is now pushing an updated and more robust version of this

tautschnig replied:
It would have worked if you used 1) `bash` as the shell and 2) had the `globstar` option set (for those versions of `bash` that even support it). As it was, `**` was not meaningful. But then you've now found a different solution.

NlightNFotis committed:
Move building of artifacts to separate subdirectory.

This avoids potential issues with clashing translation unit name in case
pre-existing translation units exist in the archiving directory.

thomasspriggs commented on src/libcprover-cpp/add_dependencies.sh:
```diff
@@ -19,6 +19,14 @@ LIB_LIST=$@
+mkdir add_dependencies_tmp
+cd add_dependencies_tmp
+
+# The full path of the current "root" directory
```
:pick: I found this comment confusing on first reading. Surely `pwd` Prints the Working Directory not the root directory which would be `/`? On second reading I guess this means the root of the working tree, rather than the root of the filing system, but the overloaded terminology makes this confusing.

thomasspriggs commented on src/libcprover-cpp/add_dependencies.sh:
```diff
@@ -33,18 +41,18 @@ for lib in ${LIB_LIST}; do
-  # Rename all object file in the library prepending "${LIBNAME}_" to avoid
-  # clashes.
+  # Rename all object files in the library prepending "${LIBNAME}_" to avoid
+  # clashes, and move to the "root" folder.
```
:pick: Mixed terminology. You used "folder" here and "directory" previously. It would be nice to choose one term and use it consistently...

TGWDB reviewed (approved):
I agree with the comments by @thomasspriggs and also would prefer not to have a `TODO` remaining. That said, they're not blockers to this progressing for me.

NlightNFotis commented:
@TGWDB The `TODO` needs to stay in for now, as it's something to remind us to check out later what the impact of this change is on the runner caches.

We can't do it now (it requires multiple rounds of CI after this PR has been merged) and it's something that we really need to keep track of going forward, as we've already blown the caches way above the limit we're allowed.
